substrate/in-instructions/primitives/src/incoming.rs

```diff
@@ -0,0 +1,52 @@
+use scale::{Encode, Decode, MaxEncodedLen};
+use scale_info::TypeInfo;
+
+#[cfg(feature = "std")]
+use serde::{Serialize, Deserialize};
+
+use sp_core::{ConstU32, bounded::BoundedVec};
+
+use serai_primitives::NativeAddress;
+
+use crate::{MAX_DATA_LEN, ExternalAddress};
+
+#[derive(Clone, Copy, PartialEq, Eq, Debug, Encode, Decode, MaxEncodedLen, TypeInfo)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+pub enum Application {
+  DEX,
+}
+
+#[derive(Clone, PartialEq, Eq, Debug, Encode, Decode, MaxEncodedLen, TypeInfo)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+pub struct ApplicationCall {
+  application: Application,
+  data: BoundedVec<u8, ConstU32<{ MAX_DATA_LEN }>>,
+}
+
+#[derive(Clone, PartialEq, Eq, Debug, Encode, Decode, MaxEncodedLen, TypeInfo)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+pub enum Target {
+  Application(ApplicationCall),
+  Address(NativeAddress),
+}
+
+#[derive(Clone, PartialEq, Eq, Debug, Encode, Decode, MaxEncodedLen, TypeInfo)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+pub struct InInstruction {
+  pub origin: ExternalAddress,
+  pub target: Target,
+}
+
+#[derive(Clone, PartialEq, Eq, Debug, Encode, Decode, MaxEncodedLen, TypeInfo)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+pub struct ExternalInInstruction {
```

Comment:
I feel like we're missing fields here. Only origin and target?

Reply:
It's because the entire call is crammed into Target.

This does also need an amount field in general, I just didn't get to that part of the pallet. Please revisit after I make the other discussed changes (EII dropped for RII).